[mod_articles_news] Option Intro/Full Images of the article

carlitorweb opened the pull request:
Pull Request for Issue #19599 , #16495

### Summary of Changes
Option to use or not, the intro image field or the full image field of the article

### Testing Instructions
1. Go to Extensions ->Modules and create a new Articles - Newsflash
2. Put Show Images option, for any of the 3 options that list. Also, configure the module to display correctly in some position of the template
3. Go to Content -> Articles, and use in one of the articles that the module will show, any of the fields "Intro image" or "Full article image"
4. Go to the frontend, and check depending on the option that you put in the module, it shows or not, the image correctly

### Expected result
The module shows or not, the fields "Intro image" or "Full article image" of the article correctly

### Actual result
The current option only display images in the article context.

### Documentation Changes Required
New module option added.

carlitorweb committed:
[mod_articles_news] Option Intro/Full Images of the article

Option to use or not, the intro image field or the full image field of
the article

Quy commented on modules/mod_articles_news/helper.php:
```diff
@@ -119,9 +119,32 @@ public static function getList(&$params)
+				{
+					$item->imageSrc = htmlspecialchars($images->image_intro, ENT_COMPAT, 'UTF-8');
+					$item->imageAlt = htmlspecialchars($images->image_intro_alt, ENT_COMPAT, 'UTF-8');
+					if ($images->image_intro_caption) 
```
Add blank line above.

Quy commented on modules/mod_articles_news/helper.php:
```diff
@@ -119,9 +119,32 @@ public static function getList(&$params)
+				{
+					$item->imageSrc = htmlspecialchars($images->image_fulltext, ENT_COMPAT, 'UTF-8');
+					$item->imageAlt = htmlspecialchars($images->image_fulltext_alt, ENT_COMPAT, 'UTF-8');
+					if ($images->image_intro_caption) 
```
Add blank line above.

Quy commented on modules/mod_articles_news/helper.php:
```diff
@@ -119,9 +119,32 @@ public static function getList(&$params)
+						$item->imageCaption = htmlspecialchars($images->image_intro_caption, ENT_COMPAT, 'UTF-8');
+					}					
+				}				
+				else if ($params->get('image') == "full" && isset($images->image_fulltext) && !empty($images->image_fulltext))
```
Change to `elseif`. Change `"` to `'`. Change `==` to `===`.

Quy commented on modules/mod_articles_news/helper.php:
```diff
@@ -119,9 +119,32 @@ public static function getList(&$params)
 			{
-				$item->introtext = preg_replace('/<img[^>]*>/', '', $item->introtext);
+				$images = json_decode($item->images);
+				$item->imageSrc = "";
```
Change `""` to `''`

Quy commented on modules/mod_articles_news/helper.php:
```diff
@@ -119,9 +119,32 @@ public static function getList(&$params)
 
-			if (!$params->get('image'))
+			// Get the data of the image, if it is going to be shown
+			if ($params->get('image') != "none")
```
Change `"` to `'`

Quy commented on modules/mod_articles_news/helper.php:
```diff
@@ -119,9 +119,32 @@ public static function getList(&$params)
+				$item->imageSrc = "";
+				$item->imageAlt = "";
+				$item->imageCaption = "";
+				if ($params->get('image') == "intro" && isset($images->image_intro) && !empty($images->image_intro))
```
Change `"` to `'`. Change `==` to `===`.

Quy commented on modules/mod_articles_news/helper.php:
```diff
@@ -119,9 +119,32 @@ public static function getList(&$params)
+					{
+						$item->imageCaption = htmlspecialchars($images->image_fulltext_caption, ENT_COMPAT, 'UTF-8');
+					}
+					
```
Remove blank line.

Quy commented on modules/mod_articles_news/mod_articles_news.xml:
```diff
@@ -46,14 +46,14 @@
 					description="MOD_ARTICLES_NEWS_FIELD_IMAGES_DESC"
-					class="btn-group btn-group-yesno"
-					default="0"
+					default="1"
```
Why `1`?

carlitorweb replied:
Was a mistake. Now is okay

Quy commented on language/en-GB/en-GB.mod_articles_news.ini:
```diff
@@ -30,6 +30,8 @@ MOD_ARTICLES_NEWS_FIELD_TRIGGEREVENTS_DESC="Triggers additional plugin events to
 MOD_ARTICLES_NEWS_FIELD_TRIGGEREVENTS_LABEL="Trigger Plugin Events"
 MOD_ARTICLES_NEWS_FIELD_SHOWINTROTEXT_DESC="Show or hide the article intro text."
 MOD_ARTICLES_NEWS_FIELD_SHOWINTROTEXT_LABEL="Show Intro Text"
+MOD_ARTICLES_NEWS_OPTION_INTROIMAGE="Intro image"
```
Change to `Image`

Quy commented on modules/mod_articles_news/tmpl/_item.php:
```diff
@@ -22,6 +22,15 @@
+	<figure class="newsflash-image">
+		<img src="<?php echo $item->imageSrc;?>" alt="<?php echo $item->imageAlt;?>">
+		<figcaption>
+			<?php echo $item->imageCaption;?>
```
Add space after `;`

Quy commented on modules/mod_articles_news/tmpl/_item.php:
```diff
@@ -22,6 +22,15 @@
 
+<?php if ($params->get('image') != "none" && !empty($item->imageSrc)) : ?>	
+	<figure class="newsflash-image">
+		<img src="<?php echo $item->imageSrc;?>" alt="<?php echo $item->imageAlt;?>">
```
Add space after `;`

Quy commented on modules/mod_articles_news/tmpl/_item.php:
```diff
@@ -22,6 +22,15 @@
 	</<?php echo $item_heading; ?>>
 <?php endif; ?>
 
+<?php if ($params->get('image') != "none" && !empty($item->imageSrc)) : ?>	
```
Change `"` to `'`

carlitorweb commented:
Is done, thank you @Quy

carlitorweb committed:
Code review

Fixed default attribute of the image field inside the xml file.
Added a condition in case the caption of the image is empty, for no show
unnecessarily the figcaption tag

carlitorweb committed:
Code review

ghost commented:
I have tested this item :white_check_mark: successfully on a3b2f089c65daac16e8ebb4f26099122700f66bd<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20169">issues.joomla.org/tracker/joomla-cms/20169</a>.</sub>

bubi-luka commented:
I have tested this item :white_check_mark: successfully on a3b2f089c65daac16e8ebb4f26099122700f66bd<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20169">issues.joomla.org/tracker/joomla-cms/20169</a>.</sub>

ghost commented:
Ready to Commit after two successful tests.

Quy commented on modules/mod_articles_news/helper.php:
```diff
@@ -119,9 +119,33 @@ public static function getList(&$params)
+				$item->imageSrc = '';
+				$item->imageAlt = '';
+				$item->imageCaption = '';
+				if ($params->get('image') === 'intro' && isset($images->image_intro) && !empty($images->image_intro))
```
Add blank line above.

Quy commented on modules/mod_articles_news/tmpl/_item.php:
```diff
@@ -22,6 +22,17 @@
+<?php if ($params->get('image') != 'none' && !empty($item->imageSrc)) : ?>	
+	<figure class="newsflash-image">
+		<img src="<?php echo $item->imageSrc; ?>" alt="<?php echo $item->imageAlt; ?>">
+		<?php if(!empty($item->imageCaption)) : ?>
```
Add space after `if`

Quy commented on modules/mod_articles_news/mod_articles_news.xml:
```diff
@@ -46,14 +46,14 @@
-					<option value="0">JNO</option>
+					<option value="intro">MOD_ARTICLES_NEWS_OPTION_INTROIMAGE</option> 
+					<option value="full">MOD_ARTICLES_NEWS_OPTION_FULLIMAGE</option>
+					<option value="null">JNO</option>
```
Change to `none`?

carlitorweb commented:
Thank you again @Quy . I need have more careful with this small things.

carlitorweb committed:
Code review

Quy commented on modules/mod_articles_news/helper.php:
```diff
@@ -119,9 +119,34 @@ public static function getList(&$params)
+				$item->imageAlt = '';
+				$item->imageCaption = '';
+				
+				if ($params->get('image') === 'intro' && isset($images->image_intro) && !empty($images->image_intro))
```
Remove `isset($images->image_intro) ` since `empty` includes `isset` check.

Quy commented on modules/mod_articles_news/helper.php:
```diff
@@ -119,9 +119,34 @@ public static function getList(&$params)
+						$item->imageCaption = htmlspecialchars($images->image_intro_caption, ENT_COMPAT, 'UTF-8');
+					}					
+				}				
+				elseif ($params->get('image') === 'full' && isset($images->image_fulltext) && !empty($images->image_fulltext))
```
Remove `isset($images->image_fulltext)` for same reason as stated above.

Quy commented on modules/mod_articles_news/helper.php:
```diff
@@ -119,9 +119,34 @@ public static function getList(&$params)
 
-			if (!$params->get('image'))
+			// Get the data of the image, if it is going to be shown
+			if ($params->get('image') != 'none')
```
Change to `!==`

Quy commented on modules/mod_articles_news/tmpl/_item.php:
```diff
@@ -22,6 +22,18 @@
 	</<?php echo $item_heading; ?>>
 <?php endif; ?>
 
+<?php if ($params->get('image') != 'none' && !empty($item->imageSrc)) : ?>	
```
Change to `!==`

Quy commented on language/en-GB/en-GB.mod_articles_news.ini:
```diff
@@ -30,6 +30,8 @@ MOD_ARTICLES_NEWS_FIELD_TRIGGEREVENTS_DESC="Triggers additional plugin events to
 MOD_ARTICLES_NEWS_FIELD_SHOWINTROTEXT_DESC="Show or hide the article intro text."
 MOD_ARTICLES_NEWS_FIELD_SHOWINTROTEXT_LABEL="Show Intro Text"
+MOD_ARTICLES_NEWS_OPTION_INTROIMAGE="Intro Image"
+MOD_ARTICLES_NEWS_OPTION_FULLIMAGE="Full Article Image"
```
Sort keys in alpha order.

Quy commented on modules/mod_articles_news/mod_articles_news.xml:
```diff
@@ -46,14 +46,14 @@
 					description="MOD_ARTICLES_NEWS_FIELD_IMAGES_DESC"
-					class="btn-group btn-group-yesno"
-					default="0"
+					default="intro"
```
This could be a b/c issue because when this value is `0`, then it will be `intro` and not `none`.

carlitorweb replied:
But how that value will be 0 ? default is "intro" and other option is "full" and "none"

Quy replied:
Lets say I edit an existing Newsflash module without changing this setting. The original values are `0` and `1` with `0` selected, but will default to `intro` after saving.

carlitorweb replied:
I got it now

Quy commented:
You're very welcome. For reference, read [Coding Standards Manual](https://developer.joomla.org/coding-standards.html).

carlitorweb committed:
Code review

Quy commented on modules/mod_articles_news/helper.php:
```diff
@@ -119,9 +119,34 @@ public static function getList(&$params)
+				$item->imageSrc = '';
+				$item->imageAlt = '';
+				$item->imageCaption = '';
+				
```
Remove tabs.

carlitorweb replied:
Sorry, I do not get this one. What tabs?

Quy replied:
See screenshot:
![tabs](https://user-images.githubusercontent.com/368084/38838150-c6ff596c-4189-11e8-9c31-e5311de58ed2.jpg)

carlitorweb replied:
got it. thank you again

Quy commented on modules/mod_articles_news/tmpl/_item.php:
```diff
@@ -22,6 +22,18 @@
+<?php if ($params->get('image') !== 'none' && !empty($item->imageSrc)) : ?>	
+	<figure class="newsflash-image">
+		<img src="<?php echo $item->imageSrc; ?>" alt="<?php echo $item->imageAlt; ?>">
+
```
Remove blank line.

carlitorweb committed:
Code review

Revert default value to 0, to avoid b/c issue

carlitorweb commented:
Done again @Quy . I revert also the "default" value to 0

carlitorweb committed:
Code review

Quy commented on modules/mod_articles_news/helper.php:
```diff
@@ -119,9 +119,34 @@ public static function getList(&$params)
+					if ($images->image_intro_caption) 
+					{
+						$item->imageCaption = htmlspecialchars($images->image_intro_caption, ENT_COMPAT, 'UTF-8');
+					}					
```
More tabs to delete.

![tabs2](https://user-images.githubusercontent.com/368084/38838717-6df25fe2-418c-11e8-850b-a88b31bcd54e.jpg)

carlitorweb committed:
Code review

brianteeman commented:
please remove the RTC status as this PR is not backwards compatible and it also changes the way that the module works

**Issue 1**
Create an article with an image INSIDE the article
create a newsflash module and set the show images to NO
Test and you will see that there are no images displayed
Apply this PR and do NOT change anything in the module
Check the module on the front end and you will see that images ARE displayed

**Issue 2**
There is no longer a way to not display images that are INSIDE the article
The no image option  only stops the display of intro and full images
the intro image option will display the images INSIDE the article and the intro image
The full image option will display the images INSIDE the article and the full image

brianteeman commented:
I have tested this item :red_circle: unsuccessfully on 9d9e6065012943b662444d3dc6a8c39b48d9caa3<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20169">issues.joomla.org/tracker/joomla-cms/20169</a>.</sub>

ghost commented:
@brianteeman Status is back on Pending by @Quy at Issue Tracker but Label "RTC" on Github is shown, can't remove it here.

brianteeman commented:
@franz-wohlkoenig if a status is changed on the issue tracker it will be updated on github but only when a comment is made. Your comment triggered the change. So in future if a change is made on the issue tracker you should always make a comment explaining that yu made the change and then it will work here on github too

ghost commented:
thanks Brian, so its vica versa working to set RTC and wrote here > haven't thought about it :-

carlitorweb commented:
@brianteeman the issue you talk about is more related to the user context than the module. Why need the module remove any image you put INSIDE the article? If I want my article to have or not an image in the introtext, then I just put it or not, INSIDE the introtext. 
Anyway, can be another param asking for this, in case the user wants also, show or not, the image INSIDE the article

carlitorweb commented:
Also, in my opinion, I think the option to put an image that belongs to the text of the article, is an option only a few users use it.

brianteeman commented:
1. Unless you have a magic ball you can not know how many people use the current options
2. You have to consider backwards compatibility. You can not make a change that will effect the display of millions of web sites

carlitorweb commented:
No, I do not have a magic ball, that is why I wrote "in my opinion", just counting with my small knowledge about this option. 
Ok, what I will do is, revert the change, and make a new param leaving the old intact. In this way, we are only adding a new option to the module, without modifying it

carlitorweb committed:
Revert change. Adding new option for the intro/full image

carlitorweb commented:
I revert the change. Now, this option is with a new param. Also, I modify the description for the old one, trying to avoid confusion, if is possible change this description.

brianteeman commented:
Thank you

carlitorweb committed:
alpha order

carlitorweb commented:
You welcome @brianteeman

brianteeman commented:
B/C images all fixed now. 

I think it would be best if you updated the old string "Show Images" to "Show article images" or something like that - just so that people dont think it is related to the other options below

carlitorweb commented:
Is true. Done. Thank you again.

carlitorweb committed:
Change old string for better descriptive label

ReLater commented:
I have tested this item :white_check_mark: successfully on 68e2e54c7494568e2565b6ff566b75b89c92732b<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20169">issues.joomla.org/tracker/joomla-cms/20169</a>.</sub>

carlitorweb committed:
Trying better label/desc string

carlitorweb committed:
Code review

ReLater commented on language/en-GB/en-GB.mod_articles_news.ini:
```diff
@@ -7,8 +7,8 @@ MOD_ARTICLES_NEWS="Articles - Newsflash"
 MOD_ARTICLES_NEWS_FIELD_CATEGORY_DESC="Select Articles from a specific Category or a set of Categories. If no selection will show all categories as default."
-MOD_ARTICLES_NEWS_FIELD_IMAGES_ARTICLE_DESC="Display article intro or full image."
-MOD_ARTICLES_NEWS_FIELD_IMAGES_ARTICLE_LABEL="Show Intro-Full Image"
+MOD_ARTICLES_NEWS_FIELD_IMAGES_ARTICLE_DESC="Display article, intro or full image."
```
@brianteeman help pls.
I don't think that the comma is correct here. The string should describe something like:
"Show the intro image or the full image that you selected in the article"
but not
"Show article image or the intro image or the full image"

(for the images inside article text we have another setting)

brianteeman replied:
sorry I thought it was display article or intro or full
in this context the word article is not needed and as you can see fro my error it can be misleading
i dont think we ever refer to them as article images any way so
why not simply say 
Display the intro or full image

carlitorweb replied:
I wrote the "article" word, trying not to confuse the one who reads it, with the other param. "Intro image" and "Full image", are well known, but include the "article" word, I think it locates better which fields we're talking about.

ReLater replied:
Somehow or other but the comma is wrong. I myself would prefer a description without "article"  here. For me an "article" image (inside the editor text) is distinct from an "intro" image is distinct from a "full" image. The description would be clearer for me as a not native English speaker.

carlitorweb committed:
Code review

carlitorweb commented:
Ok done. Hope is better now.

ReLater commented:
I have tested this item :white_check_mark: successfully on 5efd862b7e9d7a29413f322e6e101779b813a9cc<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20169">issues.joomla.org/tracker/joomla-cms/20169</a>.</sub>

Quy commented:
I have tested this item :white_check_mark: successfully on 5efd862b7e9d7a29413f322e6e101779b813a9cc<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20169">issues.joomla.org/tracker/joomla-cms/20169</a>.</sub>

Quy commented:
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20169">issues.joomla.org/tracker/joomla-cms/20169</a>.</sub>

ute-sch commented:
Please help (my Joomla's up to date):
The Article-Newsflash Module doesn't show the Intro or Full image, as offered in the field.

What I found out - sorry, I'm not a programmer: 

In the options field there's an english-german language mixup -  see the screenshot:
[http://www.schultz-hamburg.de/joomla-newsflash.png](url)
Doesn't the module find the appropriate value?
I couldn't find the file where the **values** are **translated** and assigned to the corresponding keys.

I only found the german translation file _de-DE_mod_articles-news_ini_ where the following lines are missing:
`MOD_ARTICLES_NEWS_FIELD_IMAGES_ARTICLE_DESC="... german translation ..."
MOD_ARTICLES_NEWS_FIELD_IMAGES_ARTICLE_LABEL="... german translation ..."`
This is not a big issue, it's only the translation of the label and the description - so I tried and added it on my site. Of course, it didn't help with the bug.

Thanks for your patience!

franz-wohlkoenig commented:
@ute-sch please ask on [Forum](https://forum.joomla.org/)
